text/0000-export.md

```diff
@@ -0,0 +1,403 @@
+The hash must have been randomly generated to ensure uniqueness (which is part of the unsafe promise).
+
+```rust
+#[export(unsafe_stable_abi = "ca83050b302bf0644a1417ac3fa6982a")]
+#[repr(C)]
+pub struct ListNode {
+    next: *const ListNode,
+    value: i32,
+}
+```
+
+In this case, using the type as part of a function signature will not result in a hash based on the full (recursive) type definition,
+but will instead be based on the user provided hash (and the size and alignment of the type).
+
+### Standard Library
+
+Once the ["crabi"](https://github.com/rust-lang/rust/pull/105586) feature has progressed far enough,
+we should consider adding `#[export]` attributes to some standard library types, effectively committing to a stable ABI for those.
+For example, `Option`, `NonZero`, `File`,
+and many others are good candidates for `#[export(unsafe_stable_abi)]`
+(if the "crabi" ABI doesn't already handle them specially).
+
+## Future Possibilities
+
+- A `#[no_export]` attribute, which can be useful when marking an entire crate or module as `#![export]`.
+- Options within the `#[export(…)]` attribute for e.g. a future name mangling scheme version.
+- `#[export(by_path)]` to export a symbol only based on the path, without the hash of relevant safety/type information.
+  This is useful in situations where safety is not a primary concern, to simplify cases like using Rust code from another language (with a simple symbol name).
+  (Importing (or using) such a symbol from Rust will be `unsafe`.)
+- `#[export(opaque)]` (or `#[export(indirect)]`) for opaque types that can only be used indirectly (e.g. through a pointer or reference) in exported items,
+  such that their size is not a stable ABI promise.
+- Exportable `static`s.
+- Exportable `trait`s, for e.g. dynamic trait objects. (See also https://github.com/rust-lang/rust/pull/105586.)
+- A tool to create a stripped, 'dynamic import only' version of the crate source code,
+  with only the exported items, without the function bodies. (Essentially a "header file" or "bindings-only crate".)
+  - Alternatively or in addition to that, a way to optionally include such a "header file" or "bindings"
+    (or the same metadata in some other (perhaps Rust-agnostic) format)
+    inside the resulting dynamic library file.
+- Allow exporting two identically named items to create a dynamic library that is backwards compatible with an older interface,
+  including both a symbol for the old and new interface.
+- Next to the hash of the type information,
+  additionally and optionally include the full type information in an extra section,
+  to allow for (debug) tools to accurately diagnose mismatched symbol errors.
+- Some kind of `#[export_inline]` feature to allow for functions that will be inlined into the calling crate,
+  rather than being part of the dynamic library, which will only be able to call exported items.
+
+## Alternatives
+
+- Alternatives for using a hash of all relevant type information:
```

Comment:
I think this RFC might want to talk about why this is the right choice where we previously made a different one in the move from legacy to v0 (which moved away from a hash, at least mostly). I think given the use case here this may be better, but the discussion seems worthwhile.

Reply:
it seems like a good idea to me to combine both of them -- `v0` mangling so debuggers and similar can get the full source symbol name including generic arguments, and the hash to detect ABI breaks (ABI breaks won't necessarily change the `v0` mangling, such as reordering struct fields)